drivers/adc/adc_tla2021.c

```diff
@@ -0,0 +1,272 @@
+
+	return 0;
+}
+
+int tla2021_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	int ret;
+
+	struct tla2021_data *data = dev->data;
+
+	if (seq->channels != BIT(0)) {
+		LOG_ERR("Selected channel(s) not supported: %x", seq->channels);
+		return -EINVAL;
+	}
+
+	if (seq->resolution != 12) {
+		LOG_ERR("Selected resolution not supported: %d", seq->resolution);
+		return -EINVAL;
+	}
+
+	if (seq->oversampling) {
+		LOG_ERR("Oversampling is not supported");
+		return -EINVAL;
+	}
+
+	if (seq->calibrate) {
+		LOG_ERR("Calibration is not supported");
+		return -EINVAL;
+	}
+
+	if (!seq->buffer || !seq->buffer_size) {
+		LOG_ERR("Buffer invalid");
+		return -EINVAL;
+	}
+
+	const size_t num_samples = 1 + (seq->options ? seq->options->extra_samplings : 0);
+
+	if (!seq->buffer || seq->buffer_size < (num_samples * sizeof(int16_t))) {
+		LOG_ERR("No buffer or too small");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	if (seq->options && seq->options->extra_samplings && seq->options->interval_us) {
+		k_timer_start(&data->timer, K_NO_WAIT, K_USEC(seq->options->interval_us));
+	}
+
+	enum adc_action action = ADC_ACTION_CONTINUE;
+
+	for (int i = 0; i < num_samples; i += (action == ADC_ACTION_REPEAT ? 0 : 1)) {
+		uint16_t reg = data->reg_config;
+
+		if (seq->options && seq->options->extra_samplings && seq->options->interval_us) {
+			k_event_wait_all(&data->event, EVENT_START_CONVERSION, true, K_FOREVER);
+		}
+
+		/*
+		 * Enable single-shot conversion
+		 */
+		WRITE_BIT(reg, REG_CONFIG_MODE_pos, 1);
+		WRITE_BIT(reg, REG_CONFIG_OS_pos, 1);
+		ret = write(dev, REG_CONFIG, reg);
+		if (ret) {
+			k_timer_stop(&data->timer);
+			k_mutex_unlock(&data->lock);
+			return ret;
+		}
+
+		do {
+			ret = read(dev, REG_CONFIG, &reg);
+			if (ret < 0) {
+				k_timer_stop(&data->timer);
+				k_mutex_unlock(&data->lock);
+				return ret;
+			}
+		} while (!(reg & REG_CONFIG_OS_msk));
+
+		ret = read(dev, REG_DATA, &((int16_t *)seq->buffer)[i]);
+		if (ret) {
+			k_timer_stop(&data->timer);
+			k_mutex_unlock(&data->lock);
+			return ret;
+		}
+
+		((int16_t *)seq->buffer)[i] >>= REG_DATA_pos;
+
+		if (seq->options && seq->options->callback) {
+			action = seq->options->callback(dev, seq, i);
+
+			if (action == ADC_ACTION_FINISH) {
+				break;
+			}
+		}
+	}
+
+	k_timer_stop(&data->timer);
+	k_mutex_unlock(&data->lock);
+
+	return 0;
+}
+
+void expiry_fn(struct k_timer *timer)
+{
+	struct tla2021_data *data = CONTAINER_OF(timer, struct tla2021_data, timer);
+
+	k_event_post(&data->event, EVENT_START_CONVERSION);
+}
+
+void stop_fn(struct k_timer *timer)
+{
+	struct tla2021_data *data = CONTAINER_OF(timer, struct tla2021_data, timer);
+
+	k_event_clear(&data->event, EVENT_START_CONVERSION);
+}
+
+static int tla2021_init(const struct device *dev)
+{
+	int ret;
+
+	const struct tla2021_config *config = dev->config;
+	struct tla2021_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+	k_event_init(&data->event);
```

Comment:
I have not seen this API used in any driver context. For that reason, I would guess that it's probably not the best choice.

@anangl - can you provide some advice?

Reply:
I think events are pretty new. Do you see a different way to get a somewhat stable sample rate?

Reply:
@dp7hgh7 - a lot of the sensor drivers use a thread or workqueue I would check those out. There might even be a doc about it.

I just thought I would ask the ADC maintainer (@anangl) first to see if there was a preferred way of doing things for ADC drivers.

Reply:
I see. I'll have a look, thanks!

Reply:
I reworked my solution. Actually I completely overlooked the `adc_context.h`... It's now used and very close the the adsxxxx drivers already in the tree.